README
SubC is a simple subset of C originally written by 
Nils Holms (http://www.t3x.org/) for teaching newcomers 
how to write a compiler.

This is a reimplementation of SubC based on Nils compiler 
(subc-20161212.tgz) and the Go standard packages in Go. 
The code generation is based on the SubC compiler 
while the lexer, parser, and type checker is based on the Go 
standard library packages. The code emitted by this compiler
should match exactly with what SubC generates.

The linux amd64/i386 code generator has been tested, but the Makefile
only supports linux amd64 right now.

The darwinamd64 and arm6 has not been tested and was just implemented by a transliteration.

There are some minor differences in this implementation of SubC and
the official one. The differences are listed below:

* __FILE__, __LINE__, __TIME__, __DATE__, __func__, __FUNCTION__ are supported

* goto is supported, labels can be declared after local variable declarations.

* function parameters can have no names, ie, void f(int, int a, int, char x)

* more lenient on assignments; we can assign integers to pointers after
local declarations and integer types can be assigned by pointer types.

The bugs/ directory has all the known bugs.
The subc/ dir contains the original subc code for bootstrapping
The src/ dir contains the Go source code of the compiler
The test/ dir is for test code to make sure that we generate exact same code as subc


subc/Changes
2016-12-12 Fix: char c=-2; stored -2, but char is unsigned. (also by Volkmar)
2016-12-01 Fix: f(a[1] && a[0]) would generate wrong code. Thanks, Volkmar!
2016-11-28 Fix: (a,1),2 would generate wrong code. (also by Volkmar)
2016-11-24 Fix: (x|1)|(x>1) would generate wrong code. (also by Volkmar)
2016-08-07 OpenBSD: not linking against crtbegin.o. Thanks, Volkmar Klatt!
2016-07-26 Added NetBSD/386 port supplied by Minux. Thanks!
2016-07-26 Applied signal() patch for NetBSD/x86-64. Thanks, Minux!
2016-02-17 Patch: added // comments. Thanks again, Steve!
2016-02-16 Fix: Darwin back-end emitted wrong asm code. Thanks, Steve Ward!
2016-02-16 Fix: SubC would miscompile x<256;x=0; (crash). Thanks, Quan Tran!
2016-02-16 Fix: freopen() had a precedence error. (also by Quan Tran)
2016-02-16 Fix: Constant folding miscalculated right-shift. (also by Quan Tran)
2015-02-18 Fix: Emitted wrong arguments to MULQ on x86 in struct* scaling.
2015-02-02 Fix: Long init lists in array declarations could run out of nodes.
2014-12-07 Fix: strncpy() should pad destination. Thanks, Alexey Frunze!
2014-12-07 Fix: fseek() should return 0 on success. (also by Alexey)
2014-12-07 Fix: longjmp() may not pass 0 to setjmp(). (also by Alexey)
2014-12-07 Fixed byte shift code emitted for x86. Thanks, Teran McKinney!
2014-06-22 Running extended triple test only in case of failure.
2014-06-03 Fix: syntax errors in constexprs could crash the compiler.
2014-05-25 Updated and re-organized documentation.
2014-05-24 Added operation reordering (tree skewing), e.g.: a+b*c --> b*c+a.
2014-05-24 Added strength reduction 2/2^N --> 2>>N.
2014-05-24 Added strength reduction 2*2^N --> 2<<N.
2014-05-20 Fix: convert '/' in all commands paths in Windows system().
2014-05-19 Applied patch for fixing system() on Windows. Thanks, Jean-Marc!
2014-05-18 Added OpenBSD ELF note.
2014-05-18 Fix: missed some functions when adapting OpenBSD crt0 to R-L calls.
2014-05-18 Applied more Windows patches by Jean-Marc Lienher. Thanks!
2014-05-17 Updated documentation.
2014-05-17 Changed random number test in libtest.c.
2014-05-17 Made kprintf() convert '\n' on descriptors 1,2 on DOS.
2014-05-16 Made some stdio functions emit CR,LF for '\n' on stdout/err on DOS.
2014-05-14 Added triple test script to gather data in case of failure.
2014-05-14 Added some DOS #ifdef:s to libtest.c (./file vs .\file).
2014-05-14 Adapted printf test in libtest.c for 16-bit systems.
2014-05-13 Fixed setjmp() function in DOS port (overwrote caller's memory).
2014-05-12 Updated documentation.
2014-05-10 Fix: added installation of (non-stdio) ncrt0.o to Makefile.
2014-05-10 Applied a load of Windows patches by Jean-Marc Lienher. Thanks!
2014-05-09 Fix: wrong Opt_sum_lim when BPW != sizeof(int).
2014-05-09 Removed constant expr parser, using constant folding instead.
2014-05-05 Increased NODEPOOLSZ.
2014-05-05 Windows port builds, but programs crash; back to EXPERIMENTAL.
2014-05-05 Changed winbuild.bat to print instructions, using make again.
2014-05-04 Changed triple test to use assembly output instead of binaries.
2014-05-04 Added -N (don't use stdio) flag and generation of ncrt0's.
2014-05-03 Removed "make" from winbuild.bat script (probably still broken).
2014-05-03 Fix: remove trailing '\r' in preprocessor lines.
2014-05-03 Gave up writing a signal(3) function for NetBSD. Using libc for now.
2014-05-03 Removed almost all libc dependencies from NetBSD-x86-64 crt0.
2014-05-02 Removed libc dependencies from FreeBSD-x86-64 crt0.
2014-05-01 Fixed order of argc/argv on DOS port (due to changed calling conv).
2014-05-01 Marked all untested ports "untested".
2014-05-01 Added removal of zero offsets from address computation.
2014-04-30 Adapted all remaining ports to new calling convention.
2014-04-29 Adapted NetBSD-x86-64 crt0 to new calling convention.
2014-04-29 Adapted FreeBSD-x86-64 crt0, build failed, marked BROKEN.
2014-04-28 Adapted FreeBSD-armv6 crt0 to new calling convention.
2014-04-28 Adapted FreeBSD-386 crt0 to new calling convention.
2014-04-28 Removed varargs redefinition hack (no longer needed).
2014-04-28 Changed calling convention to standard C (right-to-left).
2014-04-27 Added optimizer.
2014-04-26 Added tree abstraction.
2014-04-20 Forked this version.
2014-04-07 Fixed XSTRIP and SNM programs on 16-bit systems.
2014-04-05 Made s86 cache code for faster resolution of forward references.
2014-04-05 Fix: missing file name in error messages due to refactoring.
2014-04-03 Added DOS build script (targets/dos-8086/dosbuild.bat).
2014-04-02 Clean-up and misc. small fixes.
2014-04-01 The compiler can now be cross-compiled to DOS.
2014-04-01 Updated codumentation.
2014-04-01 Added DOS/EXE targets to Makefiles.
2014-03-31 Made compiler controller a separate executable on DOS.
2014-03-31 Factored out compiler controller.
2014-03-29 Documented bootstrapping of the DOS version.
2014-03-28 Added minimal runtime for compiling SCC86 on DOS.
2014-03-28 Added system() function for DOS.
2014-03-28 Added heap limit check to DOS _sbrk().
2014-03-27 Fix: struct_type[x] generated wrong offset.
2014-03-25 Turned __argc keyword into a local variable.
2014-03-23 Fix: #error could be mistakenly executed in negative #ifdef.
2014-03-22 Fix: fseek() could truncate files on DOS (zero-length _write()).
2014-03-22 Fix: OpenBSD sys.h file had wrong name.
2014-03-22 Changed default output file name for Windows.
2014-03-21 Fix: sizeof(void) was not an error.
2014-03-21 Made scc keep non-temporary object files.
2014-03-21 Added ufseek() (unsigned fseek) function to S86.
2014-03-20 Updated documentation.
2014-03-19 Both S86 and SLD run on 16-bit systems now.
2014-03-19 Made SLD store its relocation buffer on disk.
2014-03-19 Updated build script for windows. (untested!)
2014-03-18 Changed malloc() over-alloc. threshold for 16-bit systems.
2014-03-18 Made SLD store its relocation table on disk.
2014-03-17 Cleaned up the S86 tool chain, moved to s86/.
2014-03-17 Added setjmp() and longjmp() to DOS crt0.

subc/s86/exe.h
/*
 *	S86 -- A Simple 8086 Tool Chain
 *	EXE header format
 *	Nils M Holm, 1993,2013,2014
 *	In the public domain
 */

/* Default values */

#define	XMAGIC		"MZ"
#define XMINAL		0
#define XMAXAL		0xffff
#define XCKSUM		0
#define XENTRY		0
#define XCSEG		0
#define XOVL		0
#define XORELOC		0x1c

/* EXE header structure */

#define X_MAGIC		0	/* XMAGIC */
#define X_NBYTES	2	/* bytes in last page */
#define	X_NPAGES	4	/* number of pages */
#define X_NRELOC	6	/* number of relocation entries */
#define X_HDSIZE	8	/* size of header */
#define X_MINAL		10	/* minimum dynamic memory required */
#define X_MAXAL		12	/* max. dynamic memory */
#define X_SSEG		14	/* init stack segment */
#define X_STKPTR	16	/* init stack pointer */
#define X_CKSUM		18	/* check sum (neg. sum of all words) */
#define X_ENTRY		20	/* initial entry point (IP) */
#define X_CSEG		22	/* initial entry point (CS) */
#define X_ORELOC	24	/* offset of RELOC table */
#define X_OVLID		26	/* overlay id (0) */

/* EXE header size */

#define XHDR_SZ		28

